enhancements/machine-api/spot-instances.md

```diff
@@ -0,0 +1,359 @@
+---
+title: spot-instances
+authors:
+  - "@JoelSpeed"
+  - "@enxebre"
+reviewers:
+  - "@enxebre"
+  - "@bison"
+  - "@michaelgugino"
+approvers:
+  - "@enxebre"
+  - "@bison"
+  - "@michaelgugino"
+creation-date: 2020-02-04
+last-updated: 2020-02-04
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Spot Instances
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Enable OCP users to leverage cheaper, non-guaranteed instances to back Machine API Machines.
```

Comment:
would this be for all types of machines? (eg control and compute planes)

Reply:
Am not sure docs are enough for this. Etcd with two masters down bricks the cluster.

Reply:
agree w/ @sttts there is no valid scenario for running the control plane in its current form on spot instances.

Reply:
While I agree that there is no scenario where you would want to run the control-plane on spot instances, I think doing anything more than documenting users shouldn't do this might add a lot of extra complexity to this proposal and MAPI in general

As I understand it at the moment, the Machine API does not know what the machine it is creating is for. It get's userdata from MCO to apply to the instance when it creates it, but there's nothing in the spec of MAPI to say, "this one is a control-plane, do something special". The role of the machine is a black box to MAPI so having some check to make sure control-plane machines don't run on spot instances would require adding some extra API field which says `control-plane: true`, but then what validates that that field is actually the truth, to do that, you'd have to understand and interpret the userdata right?

Secondly, at the moment, there is no automation around control-plane machines, they are created by the installer normally, and if a user wants to add a new machine they can, but it requires several manual steps to reconfigure etcd and DNS. If we assume most people use the installer for creating control-plane machines, then, even if they did specify extra config for the spot instance creation, this wouldn't be observed by the terraform parts of the installer and as such, the control-plane machines would end up on on-demand instances anyway?

I'm relatively new to Openshift and still learning though, my understanding my be incorrect and I'm definitely open to other ideas on how to solve this problem

Reply:
Since we don't create control plane machines as spot types, I think an end user ignoring documentation and creating control plane machines that are spot types (which would require multiple, very specific actions on their part) is along the same lines of foot-shooting as deleting random instances in your AWS account.  We can't prevent users from doing super bad things.

We could detect after the fact that machines are control plane nodes, but as previously mentioned, the actual roles are determined via ignition payload from MCO, so it's not ideal to try to determine that ahead of time.

Reply:
I agree. This would be the same as protecting against setting the right value for the credentials secret, the right ignition secret, etc which we don't. Master machines are known to be pets today which let the installer set a desired input. When/if a controlPlane controller exists we could enforce opinionated validations and gating there.